src/program/command_mode.rs

```diff
@@ -0,0 +1,263 @@
+                        }
+                        input.pop();
+                        let display = format!("{}{}", cmd, input);
+                        self.render_screen(false, Some(&display))?;
+                    }
+                    Action::KeyboardInput(text) => {
+                        input.push_str(text);
+                        // Fixes ':' in command mode start
+                        if input.starts_with(cmd) {
+                            input = input[1..].to_string();
+                        }
+                        let display = format!("{}{}", cmd, input);
+                        self.render_screen(false, Some(&display))?;
+                    }
+                    Action::SwitchNormalMode => break 'command_loop,
+                    _ => continue,
+                }
+            }
+            std::thread::sleep(Duration::from_millis(1000 / 60));
+        }
+        Ok(input)
+    }
+
+    /// Enters command mode that gets user input and runs a set of possible commands based on user
+    /// input. After every command the user is set either into normal mode again or the app
+    /// terminates
+    ///
+    /// Commands:
+    ///     * ng/newglob                  requires only one addition parameter, the new current_dir.
+    ///                                     If the current image prior exists in the new glob move to that index
+    ///
+    ///     * h/help                      switches to normal mode and displays help info
+    ///
+    ///     * q/quit                      terminates the application
+    ///
+    ///     * r/reverse                   reverses current images, moving to index of current image prior to reverse
+    ///
+    ///     * df/destfolder               sets the new destination folder
+    ///
+    ///     * sort                        no argument: performs the selected sort on images, keeps
+    ///                                     moves to index of current image prior to reverse
+    ///                                   one argument: performs selected sort
+    ///
+    ///     * m/max                       set new maximum amount of images to display
+    pub fn run_command_mode(&mut self) -> Result<(), String> {
+        let input = self.get_command(":")?;
+        // after evaluating a command always exit to normal mode by default
+        self.ui_state.mode = Mode::Normal;
+        // Empty input means switch back to normal mode
+        if input.is_empty() {
+            return Ok(());
+        }
+        let input_vec: Vec<&str> = input.split_whitespace().collect();
+
+        match input_vec[0] {
+            "ng" | "newglob" => {
+                if input_vec.len() < 2 {
+                    let err_msg =
+                        String::from("Error: command \"newglob\" or \":ng\" requires a glob");
+                    return Err(err_msg);
+                }
+                let mut new_images: Vec<PathBuf>;
+                new_images = glob_path(input_vec[1])?;
```

Comment:
This would fail if the glob had a file or directory with a ` ` in it or escaped space even `\ `.

Example
`newglob /home/alex/My\ Pictures/Tour\ S*.png`

Reply:
Good catch! I completely forgot about that edge case, thanks! :+1:

What do you think about the implementation of command mode as a whole? Is there a better approach that I'm missing?

Reply:
I like this implementation better. (I mostly just want something that seems to work).

I do like that text entry doesn't depend on keycodes anymore. (Hello, international symbols 「東方」).

Does this string expand out `ng ~/*.png`?

Shellexpand may reduce some of the hand-written code for ~ expansion.
http://docs.rs/shellexpand/1.0.0/shellexpand/

Reply:
I really was thinking we'd need to do something for custom commands using

https://pest.rs/

or nom
https://github.com/Geal/nom/


Maybe eventually, though it'd be great? if we didn't have to.

Reply:
> Does this string expand out `ng ~/*.png`?
> 
> Shellexpand may reduce some of the hand-written code for ~ expansion.
> http://docs.rs/shellexpand/1.0.0/shellexpand/

As of right now that doesn't work, but I think that's more of an issue with `glob` than us. I'll look into `shellexpand` thanks :smiley:

Reply:
nice live grammar evaluator they have
Nice tools and documentation if needed though.

https://pest.rs/#editor

And a book
https://pest.rs/book/

Reply:
@Davejkane, @gurgalex potential futures issues based off of work done in command mode and discussed in #57

### Future Feature Issues
 * Infobar is easier to update/decoupled from the rest of rendering or refractored in a way where smaller components of rendering can be updated or rendered without rerendering everything
 * **Good First Issue** Info bar accepts the input of error messages, changes the current image index field to a red color and displays "Error" and then the error message in the current image path field. Update errors to now, if possible, be sent to the infobar removing "Error:" when necessary from the message
* **Good First Issue** Add the `add` command to command mode, which should follow the format `:add glob file directory` or `:a glob file directory` and append images to `self.paths.images`
* **Good First Issue** Add the `delete` command to command mode, which should follow the format `:delete /path/to/image regex` or `:d /path/to/image` otherwise delete current image from `self.paths.images`. *Don't actually delete the file*

Any other ideas? I feel like these are good first issues that require some work, but not like a crazy amount where they need to understand how the app functions as a whole.

Reply:
So I've looked into http://docs.rs/shellexpand/1.0.0/shellexpand/ and it uses [std::env::home_dir](https://doc.rust-lang.org/std/env/fn.home_dir.html) which is deprecated, Opened an [issue](https://github.com/netvl/shellexpand/issues/3). Afterwards it may be useful though!